addons/mail/static/src/messaging/entity/entity/field.js

```diff
@@ -0,0 +1,796 @@
+            return;
+        }
+        if (this.related) {
+            const [relationName, relatedFieldName] = this.related.split('.');
+            const relationField = entity.constructor.fields[relationName];
+            if (['one2many', 'many2many'].includes(relationField.relationType)) {
+                const newVal = [];
+                for (const otherEntity of entity[relationName]) {
+                    const otherField = otherEntity.constructor.fields[relatedFieldName];
+                    const v = otherField.get(otherEntity);
+                    if (v) {
+                        newVal.push(v);
+                    }
+                }
+                if (this.fieldType === 'relation') {
+                    this.set(entity, [['replace', newVal]]);
+                } else {
+                    this.set(entity, newVal);
+                }
+            } else {
+                const otherEntity = entity[relationName];
+                if (otherEntity) {
+                    const otherField = otherEntity.constructor.fields[relatedFieldName];
+                    const newVal = otherField.get(otherEntity);
+                    if (this.fieldType === 'relation') {
+                        if (newVal) {
+                            this.set(entity, [['replace', newVal]]);
+                        } else {
+                            this.set(entity, [['unlink-all']]);
+                        }
+                    } else {
+                        this.set(entity, newVal);
+                    }
+                }
+            }
+            return;
+        }
+        throw new Error("No compute method defined on this field definition");
+    }
+
+    /**
+     * Get the value associated to this field. Relations must convert entity
+     * local ids to entities instances.
+     *
+     * @param {mail.messaging.entity.Entity} entity
+     * @returns {any}
+     */
+    get(entity) {
+        if (this.fieldType === 'attribute') {
+            return this.read(entity);
+        }
+        if (this.fieldType === 'relation') {
+            const OtherEntity = entity.env.entities[this.to];
+            if (['one2one', 'many2one'].includes(this.relationType)) {
+                return OtherEntity.get(this.read(entity));
+            }
+            return this.read(entity)
+                .map(localId => OtherEntity.get(localId))
+                /**
+                 * FIXME: Stored relation may still contain
+                 * outdated entities.
+                 */
+                .filter(entity => !!entity);
+        }
+        throw new Error(`cannot get field with unsupported type ${this.fieldType}.`);
+    }
+
+    /**
+     * Get the raw value associated to this field. For relations, this means
+     * the local id or list of local ids of entities in this relational field.
+     *
+     * @param {mail.messaging.entity.Entity} entity
+     * @returns {any}
+     */
+    read(entity) {
```

Comment:
`.get()` is the high-level "read" of a field value, which notably return entities when accessing a relation of an entity. `.read()` is the low-level "read" of a field value, which must be used exclusively as internal code of entity fields.

This is the same idea for `.set()` (high-level "write") and `.write()` (low-level "write").

Reason why low-level functions are not private is because internal handling of `.set()` requires low-level read/write on other entity fields. Maybe we should mark them as "internal functions" with dunder prefix (`__`)?